tools/pkg/engine_build_configs/lib/src/build_config.dart

```diff
@@ -0,0 +1,820 @@
+  final List<GlobalBuild> builds;
+
+  /// A list of tests. The tests may have dependencies on one or more of the
+  /// builds.
+  final List<GlobalTest> tests;
+
+  /// A list of generator tasks that produce additional artifacts, which may
+  /// depend on the output of one or more builds.
+  final List<TestTask> generators;
+
+  /// A description of the upload instructions for the artifacts produced by
+  /// the global generators.
+  final List<GlobalArchive> archives;
+
+  @override
+  List<String> check(String path) {
+    final List<String> errors = <String>[];
+    errors.addAll(super.check(path));
+    for (int i = 0; i < builds.length; i++) {
+      final GlobalBuild build = builds[i];
+      errors.addAll(build.check('$path/builds[$i]'));
+    }
+    for (int i = 0; i < tests.length; i++) {
+      final GlobalTest test = tests[i];
+      errors.addAll(test.check('$path/tests[$i]'));
+    }
+    for (int i = 0; i < generators.length; i++) {
+      final TestTask task = generators[i];
+      errors.addAll(task.check('$path/generators/tasks[$i]'));
+    }
+    for (int i = 0; i < archives.length; i++) {
+      final GlobalArchive archive = archives[i];
+      errors.addAll(archive.check('$path/archives[$i]'));
+    }
+    return errors;
+  }
+}
+
+/// A "build" is a dictionary with a gn command, a ninja command, zero or more
+/// generator commands, zero or more local tests, zero or more local generators
+/// and zero or more output artifacts.
+///
+/// "builds" contains a list of maps with fields like:
+/// {
+///   "name": "",
+///   "gn": [""],
+///   "ninja": {},
+///   "tests": [],
+///   "generators": {
+///     "tasks": []
+///   }, (optional)
+///   "archives": [],
+///   "drone_dimensions": [""],
+///   "gclient_variables": {}
+/// }
+final class GlobalBuild extends BuildConfigBase {
+  factory GlobalBuild.fromJson(Map<String, Object?> map) {
+    final List<String> errors = <String>[];
+    final String? name = stringOfJson(map, 'name', errors);
+    final List<String>? gn = stringListOfJson(map, 'gn', errors);
+    final List<BuildTest>? tests = objListOfJson(
+      map, 'tests', errors, BuildTest.fromJson,
+    );
+    final List<BuildArchive>? archives = objListOfJson(
+      map, 'archives', errors, BuildArchive.fromJson,
+    );
+    final List<String>? droneDimensions = stringListOfJson(
+      map, 'drone_dimensions', errors,
+    );
+
+    late final BuildNinja? ninja;
+    if (map['ninja'] == null) {
+      ninja = BuildNinja.nop();
+    } else if (map['ninja'] is! Map<String, Object?>) {
+      ninja = null;
+    } else {
+      ninja = BuildNinja.fromJson(map['ninja']! as Map<String, Object?>);
+    }
+    if (ninja == null) {
+      errors.add('"ninja" field is malformed.');
+    }
+
+    late final List<BuildTask>? generators;
+    if (map['generators'] == null) {
+      generators = <BuildTask>[];
+    } else if (map['generators'] is! Map<String, Object?>) {
+      errors.add('"generators" field is malformed.');
```

Comment:
Is it worth perhaps adding the runtime type of like "expected: Map, actual: bool" to make debugging easier (since this is a debugging/testing tool)?

If so maybe pull out a helper function too to re-use with 'ninja' and  related.

Reply:
Good idea. Pulled out a helper `appendTypeError`.